### PR TITLE
fix(Authentication): Unable to login if 2 mail addresses on LDAP

### PR DESCRIPTION
**Pull request by dmyios**
## Description

Fixed issue where it was impossible to login if there were at least 2 mails on LDAP.

**Fixes** # MON-33148

## Type of change

- [x] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 22.10.x
- [ ] 23.04.x
- [ ] 23.10.x
- [ ] 24.04.x
- [x] master

#### Community contributors & Centreon team

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

**Review by nfragnet — CHANGES_REQUESTED** [review]
So when there's only one email address we use `trim` on it, but when there's multiple entries and we select the first we don't?

I think something is wrong here.

When you add this condition you may want to check if there's another ldap email which is is available and valid.

Also, nothing prevents `$userEmail` to be anything else than a non-empty string at the end, you should probably add a check to avoid this and raise a proper exception in this very case.

**Comment by dmyios** [comment]
> So when there's only one email address we use `trim` on it, but when there's multiple entries and we select the first we don't?
> 
> I think something is wrong here.
> 
> When you add this condition you may want to check if there's another ldap email which is is available and valid.
> 
> Also, nothing prevents `$userEmail` to be anything else than a non-empty string at the end, you should probably add a check to avoid this and raise a proper exception in this very case.

File: centreon/www/class/centreonAuth.LDAP.class.php [hunk 1]
```diff
@@ -225,18 +225,32 @@ public function updateUserDn()
...
+                        ) {
+                            $userEmail = $userInfos[$this->ldap->getAttrName('user', 'email')][0];
+                        }
+                    } elseif (
+                        is_string($userInfos[$this->ldap->getAttrName('user', 'email')])
+                        && trim($userInfos[$this->ldap->getAttrName('user', 'email')]) !== ''
+                    ) {
+                        $userEmail = $userInfos[$this->ldap->getAttrName('user', 'email')];
+                    } else {
+                        throw new Exception('User email must be a string or an array of strings.');
```
Comment (nfragnet) [inline centreon/www/class/centreonAuth.LDAP.class.php:244]: Simple question: according to the possible types value, wouldn't it be ok to have this instead (simpler to read)

```suggestion
                    $ldapEmailValue = $userInfos[$this->ldap->getAttrName('user', 'email')];
                    $userEmail = (trim(is_array($ldapEmailValue) ? current($ldapEmailValue): $ldapEmailValue));

                    if ($userEmail === '') {
                        throw new Exception('User email must be a string or an array of strings.');
```

Comment (dmyios) [reply centreon/www/class/centreonAuth.LDAP.class.php]: applied in separate commit

File: centreon/www/class/centreonAuth.LDAP.class.php [hunk 2]
```diff
@@ -254,18 +244,22 @@ public function updateUserDn(): bool
...
+                    throw new Exception('User pager must be a non-empty string');
                 }
+            } catch (Exception $ex) {
+                $this->CentreonLog->insertLog(
+                    3,
+                    'LDAP AUTH - Error : Invalid user pager : ' . $ex->getMessage()
+                );
+
+                return false;
             }
```
Comment (nfragnet) [inline centreon/www/class/centreonAuth.LDAP.class.php:263]: My 2cts
- try catch block seems a bit overkill and add unnecessary cognitive complexity for the single Exception you throw yourself IMHO
- an enhanced logger has been provided: please take the opportunity when you change some code to use it

```suggestion
            $ldapUserPager = $userInfos[$this->ldap->getAttrName('user', 'pager')];
            if (isset($ldapUserPager)) {
                $userPager = (trim(is_array($ldapUserPager) ? current($ldapUserPager) : $ldapUserPager));
            }

            if ($userPager === '') {
                CentreonLog::create()->error(
                    logTypeId: CentreonLog::TYPE_LDAP,
                    message: 'LDAP AUTH - Error : Invalid user pager : User pager must be a non-empty string',
                );

                return false;
            }
```

The same goes for the similar change above.
@Drago37 WDYT?

Comment (Drago37) [reply centreon/www/class/centreonAuth.LDAP.class.php:263]: Beware, here it's CentreonUserLog that is used, not CentreonLog :) I think we would deprecate CentreonUserLog to use CentreonLog. No ? What are the differences ? I prefere to start a request change to discuss before merging the PR.

**Review by Drago37 — CHANGES_REQUESTED** [review]

File: centreon/www/class/centreonAuth.LDAP.class.php (hunk 2 above)
Comment (Drago37) [reply centreon/www/class/centreonAuth.LDAP.class.php]: It's ok for me ;) seen together !

File: centreon/www/class/centreonAuth.LDAP.class.php [hunk 3]
```diff
@@ -225,33 +225,33 @@ public function updateUserDn()
...
+            $ldapUserPager = $userInfos[$this->ldap->getAttrName('user', 'pager')];
+            if (isset($ldapUserPager)) {
+                $userPager = (trim(is_array($ldapUserPager) ? current($ldapUserPager) : $ldapUserPager));
+            }
+
+            if ($userPager === '') {
+                CentreonLog::create()->error(
+                    logTypeId: CentreonLog::TYPE_LDAP,
+                    message: 'LDAP AUTH - Error : Invalid user pager : User pager must be a non-empty string',
+                );
```
Comment (Drago37) [inline centreon/www/class/centreonAuth.LDAP.class.php:252]: With the contact_id int the log to have more informations ;)
```suggestion
                CentreonLog::create()->error(
                    logTypeId: CentreonLog::TYPE_LDAP,
                    message: 'LDAP AUTH - Error : Invalid user pager : User pager must be a non-empty string',
                    customContext: ['contact_id' => $this->contactInfos['contact_id']]
                );
```

**Review by callapa — APPROVED** [review]

File: centreon/www/class/centreonAuth.LDAP.class.php [hunk 4]
```diff
@@ -225,33 +225,35 @@ public function updateUserDn()
...
-                    // Get the first if there are multiple entries
-                    if ($userInfos[$this->ldap->getAttrName('user', 'email')][0]) {
-                        $userEmail = $userInfos[$this->ldap->getAttrName('user', 'email')][0];
-                    }
-                } elseif ($userInfos[$this->ldap->getAttrName('user', 'email')]) {
-                    $userEmail = $userInfos[$this->ldap->getAttrName('user', 'email')];
-                }
+            $ldapEmailValue = $userInfos[$this->ldap->getAttrName('user', 'email')];
+            if (isset($ldapEmailValue)) {
+                $userEmail = (trim(is_array($ldapEmailValue) ? current($ldapEmailValue) : $ldapEmailValue));
```
Comment (callapa) [inline centreon/www/class/centreonAuth.LDAP.class.php:230]: There's no guarantee that the table isn't empty
```suggestion
                $userEmail = (trim(is_array($ldapEmailValue) && $ldapEmailValue !== [] ? current($ldapEmailValue) : $ldapEmailValue));
```

File: centreon/www/class/centreonAuth.LDAP.class.php [hunk 5]
```diff
@@ -225,33 +225,35 @@ public function updateUserDn()
...
-                    // Get the first if there are multiple entries
-                    if ($userInfos[$this->ldap->getAttrName('user', 'pager')][0]) {
-                        $userPager = $userInfos[$this->ldap->getAttrName('user', 'pager')][0];
-                    }
-                } elseif ($userInfos[$this->ldap->getAttrName('user', 'pager')]) {
-                    $userPager = $userInfos[$this->ldap->getAttrName('user', 'pager')];
-                }
+            $ldapUserPager = $userInfos[$this->ldap->getAttrName('user', 'pager')];
+            if (isset($ldapUserPager)) {
+                $userPager = (trim(is_array($ldapUserPager) ? current($ldapUserPager) : $ldapUserPager));
```
Comment (callapa) [inline centreon/www/class/centreonAuth.LDAP.class.php:246]: There's no guarantee that the table isn't empty
```suggestion
                $userPager = (trim(is_array($ldapUserPager) && $ldapUserPager) !== [] ? current($ldapUserPager) : $ldapUserPager));
```

**Review by callapa — CHANGES_REQUESTED** [review]

**Review by Drago37 — APPROVED** [review]

**Review by callapa — APPROVED** [review]

**Review by nfragnet — APPROVED** [review]

**Pull request merged**